ghcide/src/Development/IDE/Types/HscEnvEq.hs

```diff
@@ -112,10 +123,102 @@ newHscEnvEqWithImportPaths envImportPaths hscEnv deps = do
           (evaluate . force . Just $ listVisibleModuleNames hscEnv)
 
     return HscEnvEq{..}
+    where
+        indexDependencyHieFiles :: IO ()
+        indexDependencyHieFiles = do
+            dotHlsDirExists <- maybe (pure False) doesDirectoryExist mHlsDir
+            unless dotHlsDirExists deleteMissingDependencySources
+            void $ Map.traverseWithKey indexPackageHieFiles packagesWithModules
+        mHlsDir :: Maybe FilePath
+        mHlsDir = do
+            projectDir <- resRootPath =<< lspEnv se
+            pure $ projectDir </> ".hls"
+        deleteMissingDependencySources :: IO ()
+        deleteMissingDependencySources = do
+            completionToken <- newEmptyMVar
+            atomically $ unGetTQueue (indexQueue $ hiedbWriter se) $
+                \withHieDb -> withHieDb $ \db -> do
+                    removeDependencySrcFiles db
+                    putMVar completionToken ()
+            readMVar completionToken
+        indexPackageHieFiles :: Package -> [Module] -> IO ()
+        indexPackageHieFiles (Package package) modules = do
+            let pkgLibDir :: FilePath
+                pkgLibDir = case unitLibraryDirs package of
+                  [] -> ""
+                  (libraryDir : _) -> libraryDir
+                hieDir :: FilePath
+                hieDir = pkgLibDir </> "extra-compilation-artifacts"
+            modIfaces <- mapMaybeM loadModIface modules
+            traverse_ (indexModuleHieFile hieDir) modIfaces
+        indexModuleHieFile :: FilePath -> ModIface -> IO ()
+        indexModuleHieFile hieDir modIface = do
+            let hiePath :: NormalizedFilePath
+                hiePath = toNormalizedFilePath' $
+                  hieDir </> toFilePath (moduleName $ mi_module modIface) ++ ".hie"
+            hieCheck <- checkHieFile recorder se "newHscEnvEqWithImportPaths" hiePath
```

Comment:
we might want to be less granular with the `HieAlreadyIndexed` check - each package can contain hundreds/thousands of modules, and we don't want to check all these modules on every startup.

It's also wasteful because once we successfully index a whole unit/package we can assume that it doesn't change. So maybe we need to only check if a package has been indexed rather than checking for each file in the package.

This might require another table in `hiedb` to keep track of the packages we have indexed.